### DISCUSSION TITLE
Added closed exploded letters into design (R12 and closed polylines)

### DISCUSSION THREAD

**Discussion by XiaoPhysLab** [mozman/ezdxf#993] [Q&A] [answered]
Hello!

I am trying to use ezdxf to create a silicon photolithography mask, and the tool requires both the R12 format and all lines to be closed. This means that the text that I want in the file needs to be exploded. I have tried many things, and am just not able to figure this last step out. The closest I have gotten was this:
```python
import ezdxf
from ezdxf.fonts import fonts
from ezdxf import path
from ezdxf.addons import text2path

import logging
logging.basicConfig(level=logging.INFO)

doc = ezdxf.new(dxfversion=ezdxf.DXF12)
msp = doc.modelspace()

ff = fonts.FontFace(family="Courier New")
path_object = text2path.make_path_from_str("Hello World!", ff, size=100)
path.render_splines_and_polylines(msp, path_object)

doc.saveas("textpath.dxf")
```

which seems to work except that when I set the doc to R12 the letters are messed up and the shapes are not closed polylines.

Any help would be greatly appreciated!

**Comment by mozman** [comment]
Hi!

Please, always provide a workable file.

You have noticed the missing parts and it looks like the round stuff is missing:
![image](https://github.com/mozman/ezdxf/assets/4230966/b29654c8-bfe8-4987-b599-92e3a694e9a0)

The reason is that DXF R12 doesn't support the `SPLINE` entity and the function `path.render_splines_and_polylines()` bypasses the DXF version check.

You can create a DXF in version R2000 or newer and convert the DXF file by the ODA file converter to DXF R12, which gives the best result or you use the included `r12export` add-on of `ezdxf` to export any DXF version as DXF R12 - but with limitations. In both cases the `SPLINE` entities are rendered as approximated polylines:

```python
import ezdxf
from ezdxf.fonts import fonts
from ezdxf import path
from ezdxf.addons import text2path
from ezdxf.addons import r12export

doc = ezdxf.new()
msp = doc.modelspace()

ff = fonts.FontFace(family="Courier New")
path_object = text2path.make_path_from_str("Hello World!", ff, size=100)
path.render_splines_and_polylines(msp, path_object)

r12export.saveas(doc, "textpath_r12.dxf")
```

![image](https://github.com/mozman/ezdxf/assets/4230966/a27086c7-b938-402c-b8f6-f441ffb9a233)

**Reply by XiaoPhysLab** [reply]
Sorry about the working code, I forgot my manners. That seems to have fixed the words and they look great! The other problem is that the shapes need to be closed so that the device knows what is inside and what is outside. Right now for example the 0 is made up of 2 unrelated circles, but it needs to be a closed shape.

It currently looks like this:

![image](https://github.com/mozman/ezdxf/assets/136394987/84b255b8-b46c-4104-b29e-137c85f0892c)

And I believe it needs to look like this (This is the way AutoCAD explodes)

![image](https://github.com/mozman/ezdxf/assets/136394987/8723be50-e5cd-4365-ba98-efc88bb59ff6)

Any suggestions?

**Comment by mozman** [comment, marked as answer]
I have good and bad news for you:

1. good: You don't need the `path.render_splines_and_polylines()` function and therefore no `SPLINE` entities and `r12export` module
2. good: `ezdxf` has the tools to group the paths of a letter into a single polyline
3. bad: the result is not what you need

```python
from string import ascii_letters
import ezdxf
from ezdxf.fonts import fonts
from ezdxf import path
from ezdxf.addons import text2path

doc = ezdxf.new(dxfversion=ezdxf.DXF12)
msp = doc.modelspace()

ff = fonts.FontFace(family="JetBrains Mono")
text_path = text2path.make_path_from_str(ascii_letters + "0§$%&", ff, size=100)
for char in path.group_paths(path.single_paths([text_path])):
    vertices = []
    for subpath in char:
        # each subpath is a closed loop!
        # the end point and the start point of consecutive paths are connected
        vertices.extend(subpath.flattening(distance=1))
    msp.add_polyline2d(vertices)

doc.saveas("textpath_r12.dxf")
```
![image](https://github.com/mozman/ezdxf/assets/4230966/f334a69f-2daa-4e3d-8962-11534fb98857)

![image](https://github.com/mozman/ezdxf/assets/4230966/17247660-5c88-4399-ac6e-bc0829d1042b)

**Reply by XiaoPhysLab** [reply]
Thank you! This might be exactly what I need!
